PITCHME.md

```diff
@@ -0,0 +1,530 @@
+- <span style="color:green">GOOD</span> : modular interactive runtime (ROS API)
+- <span style="color:red">BAD</span> : not easy to compose, easy to shoot yourself
+- <span style="color:green">GOOD</span> : integrating with your package manager
+- <span style="color:red">BAD</span> : not quite integrated, it goes into `/opt/ros/<rosdistro>`
+
++++
+
+ROS1 Dependencies
+-----------------
+
+`rosdep resolve depkey`
+
+- enable cross-platform package management
+- <span style="color:green">GOOD</span> : can edit github's `ros/rosdistro` repository to add dependencies 
+- <span style="color:red">BAD</span> : currently can confuse ROS packages and dependencies (BUG)
+
++++
+
+ROS1 Nodes
+----------
+
+`rospy.init_node()` or `ros::init()`
+- declares the current process as a ROS node
+- <span style="color:green">GOOD</span> : one node match one process
+- <span style="color:red">BAD</span> : special `rosmaster`, `rosout` nodes
+- <span style="color:red">BAD</span> : pollute your process global state
+- <span style="color:red">BAD</span> : async programming can be VERY confusing
+C++ has nodelets, which means that one process has N node(let)s
+
++++
+
+ROS1 Launchers
+--------------
+
+`roslaunch my_pkg my_nodes.launch`
+
+- provides a centralised launch point for multiple process/nodes
+- <span style="color:green">GOOD</span> : can load yaml files to set parameters for the system
+- <span style="color:green">GOOD</span> : a launch file can include another
+- <span style="color:red">BAD</span> : namespaces logic and remapping is not as simple as it could 
+- <span style="color:red">BAD</span> : XML, but more programming-style logic is needed
+
++++
+
+ROS1 Parameters
+---------------
+
+`$ rosparam get` or `rospy.get_param()` or `ros::NodeHandle::getParam()`
+
+- get/set parameters for the running ROS system
+- <span style="color:green">GOOD</span> : centralised structured data
+- <span style="color:green">GOOD</span> : dynamically configurable 
+
++++
+
+ROS1 Messages
+-------------
+
+`$ rosmsg show`
+
+- defines message structure
+- <span style="color:green">GOOD</span> : provide an API other can use
+- <span style="color:green">GOOD</span> : provides serialization/deserialization crossplatform
+- <span style="color:red">BAD</span> : doesnt always check and validate the data type
+
++++
+
+ROS1 Topics
+-----------
+
+`rospy.Publisher()` or `ros::Publisher`
+
+`rospy.Subscriber()` or `ros::Subscriber`
+
+- allow ROSTCP/ROSUDP Push communication / event notifications
+- <span style="color:green">GOOD</span> : low overhead (binary serialization)
+- <span style="color:red">BAD</span> : lossy
+- <span style="color:red">BAD</span> : Inverted control flow (deal with your callback threadS)
+- <span style="color:red">BAD</span> : Cannot be used to interface with non-ROS systems
+- <span style="color:green">GOOD</span> : need 2 to implement pull communication
```

Comment:
Good?

Reply:
I was thinking about communication patterns: implementing a service style communication with topic is doable and can simplify your system (by using only one communication implementation underneath)  : http://wiki.ros.org/rocon_python_comms/Tutorials/indigo/What%20are%20Service%20Pairs%3F